### ISSUE TITLE
Questions about "options" fields in J and input to the IMP

### ISSUE THREAD

**Issue by grondo** [flux-framework/rfc#113]
Moving post-merge discussion from #108 here:
Below text from @morrone:

>I also made a perhaps controversial addition to both J and the IMP's input -- J gets a set of "options for the IMP" set by the requesting user, and "input to the IMP" gets a similar "options" field set by the resource owner. Something like this will be necessary to allow IMP plugins with optional per-job behavior (e.g. like spank plugins in Slurm). The user-specified options need to go in J and not the jobspec in J because we want to avoid parsing jobspec in the IMP if at all possible.

Can you expand on what use cases there might be for the user talking to the IMP? I thought we wanted to avoid that, and have only the parent instance use the new r-local be to be the only control of containers creation. I would think that a normal user would actually put system options in the attributes->system area of the jobspec, and then if the scheduler (or other part of the execution engine in the same instance) decided it needed some extra plugin functionality from the IMP to satisfy the request, it would put something in r, and the execution engine would in turn put that thing in r-local.

I didn't think that the IMP would even look at most of the fields of the J. Instead those fields are largly specific to a particular security plugin. Different security plugins might have different metadata requirements before applying the cryptographic signature. They might use different envelope formats too.

If there are good use cases, then it would be nice to hear what the user interface is going to look like for submitting a jobspec, because it sounds a little complicated to no longer have the jobspec be a nice self-contained, reusable job request. Putting some parts to replicate a job run in jobspec and others in the J envelope. Parts of the J envelope cannot be generated ahead of time...so users would need to remember the jobspec and a partial J.

**Comment by grondo** [comment]
> Can you expand on what use cases there might be for the user talking to the IMP? 

The user doesn't talk to the IMP, the user talks to site-defined plugins to the IMP. The use case is to support similar functionality we have with spank plugins which allow arbitrary command line options to be added to submission and execution commands and propagates the option arguments to remote plugins that run at various stages of execution.

However, what I proposed was a first cut at how I thought that could work.

> I would think that a normal user would actually put system options in the attributes->system area of the jobspec, and then if the scheduler (or other part of the execution engine in the same instance) decided it needed some extra plugin functionality from the IMP to satisfy the request, it would put something in r, and the execution engine would in turn put that thing in r-local.

This could work too. However, it would mean that the scheduler would need to know about all the site plugins, so that it could propagate them to the IMP's input, so it would complicate the system a bit. I was trying to keep things as simple as possible for now.

Also, it makes me a bit uneasy to let the scheduler (instance user) propagate user options to the IMP outside of *J*, essentially telling IMP plugins to run things on behalf of the user with no proof the user asked for any of it.

> If there are good use cases, then it would be nice to hear what the user interface is going to look like for submitting a jobspec, because it sounds a little complicated to no longer have the jobspec be a nice self-contained, reusable job request.

The jobspec is most reusable if it doesn't have site-specific or per-run specific system options encoded in it. So by moving those kinds of options into the "submission envelope" around jobspec we may improve reusability. I was thinking of jobspec as the reusable job request, and the extra options appended to *J* as submission-specific options.

Now that I'm thinking about it though, with the flexibility of jobspec, better resource representation in Flux, etc, I'm wondering if there might be much better ways to implement spank-like functionality in Flux, and these IMP options in *J*  would only be used as a last resort anyway. Especially for the user options I was talking about, I'd be fine leaving those out of *J* for now until there is a perceived need.

**Comment by morrone** [comment]
I think this might be a little to far to go to just avoid parsing yaml. :)  Keep in mind that consumers of the jobspec can largely ignore the sections that they don't care about.  The full jobspec parser isn't necessarily required if all that is needed is to retrieve some key/values from a mapping.

What kind of things would be site, specific, and capable of being set without changing the jobspec, but required to be changable by the user directly?  I'm not saying they don't exist, I'm just curious what sorts of things those would be.  Ditto per-run.

Could that sort of thing not be done in the job shell rather than the IMP?

**Comment by grondo** [comment]
> Keep in mind that consumers of the jobspec can largely ignore the sections that they don't care about.

Ok, you've convinced me. I should have said we're trying to keep the IMP from being dependent on any libjobspec version from flux-core, but you're right that it won't be a dependency if jobspec is just yaml.
Thanks, I'll remove the options section from *J*, and keep it as jobspec+security envelope.

> What kind of things would be site, specific, and capable of being set without changing the jobspec, but required to be changable by the user directly? I'm not saying they don't exist, I'm just curious what sorts of things those would be. Ditto per-run.

Here's a sample of what we use today on our systems. I'm not saying that these things can't be set in the jobspec. Especially using the `system` section as you suggest, they probably can. However, while trying to get to S4 as quickly as possible, I think I fell into the trap of mimicking current behavior very closely. If the IMP can parse a systems:options section I believe we're covered, however I'm still including some of our currently used plugins here since they are good use cases to think about

 * drop-caches.lua: Adds `--drop-caches=[pagecache|slab]` option to srun/sbatch/etc which writes to `/proc/sys/vm/drop_caches` on user's behalf.
 * hugepages.lua: Adds `--hugepages=N[KMG]` which attemps to create hugepages available on a node for the duration of a user's job
 * overcommit-memory.lua: Adds option to choose overcommit mode for node for duration of user's job
 * thp.lua: Adds `--thp=[always|never]` allowing user to control transparent hugepage support on nodes of job
 * ip-isolate.lua: Adds an `--ip-isolate` option to srun/sbatch/salloc to allow users to configure whether nodes in their job can talk to non-job job nodes (over IP)
 * mpibind.so: LLNL's custom automatic best-guess CPU affinity for tasks of a job, with options to disable or restrict binding
 * renice.so: change nice priority of tasks of job (not sure if this one is actually used, it is the demo plugin for spank)
 * use-env.so: Set and/or modify environment variables using site and user configuration file, with named environments. 

Other sites also have custom plugins, e.g. CEA's [x11](https://github.com/hautreux/slurm-spank-x11) plugin, etc.

So, there is definitely a requirement for *something* to provide this functionality in S4 timeframe.

**Comment by grondo** [comment]
> Ditto per-run.

For many of the examples above, users may want to modify behavior per-run or per-submission. An obvious example would be for `--drop-caches`. Users may want to drop the pagecache for test run or some other kind of benchmark, but normally they would not want to touch the caches. I'm not sure two different jobspecs for this case is the best idea, since the two versions could easily get out of sync in other ways.

You could mitigate this by allowing submission tool to modify an `options` section of the jobspec, which would be fine with me. As a side-effect, the submitted jobspec is not really reusable since there may be a different document per submission. That is probably ok though.

**Comment by morrone** [comment]
> however I'm still including some of our currently used plugins here since they are good use cases to think about

Thanks!  That is a lot clearer to me now.  Some of those might be doable in the jobshell, but a lot of them will need root so I see why you want to do them in the IMP.

**Comment by grondo** [comment]
Thanks. Sorry I didn't pull those examples out sooner.

On Oct 10, 2017 9:36 AM, "Christopher J. Morrone" <notifications@github.com>
wrote:

however I'm still including some of our currently used plugins here since
they are good use cases to think about

Thanks! That is a lot clearer to me now. Some of those might be doable in
the jobshell, but a lot of them will need root so I see why you want to do
them in the IMP.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/flux-framework/rfc/issues/113#issuecomment-335533081>,
or mute the thread

**Comment by grondo** [comment]
@morrone, what say you about the path to the job shell that is currently included in *J*?
This path is almost like a user provided option to the IMP (if missing use the IMP configured default), so the *job shell path* is very similar to the system options we're moving into jobspec.

I'm on the fence on this one. The path to the job shell is probably quite system/site dependent, and I can see a case where a user would want to select at runtime, though as we already said this is not an issue if the system/options section of jobspec can be rewritten at submit time. We could also make this one go away and just say that there is only one supported system job shell (selected by the IMP) for multi-user jobs (at least for the S4 time frame).

**Comment by garlick** [comment]
It would be nice if users (e.g. DEG) could at least test new job shells on a per job basis without sys admin intervention.  (I can imagine Adam Moody et al having fun with that one)

**Comment by grondo** [comment]
Yeah, that's the main use case I see. However, you *could* accomplish the same thing by running a new single user instance, but I still see value in per-job user-selectable job shell.

**Comment by garlick** [comment]
Oh right, the new instance is a way around that.

Sorry if I've forgotten how this works and this is off topic, but how does the user specify debug options, for example to run under totalview?  The ptrace initial sync would be implemented in the job shell correct?

**Comment by grondo** [comment]
Yeah, that's another good example of a case where options in a "baseline" jobspec would need to be modified at submission or run time.

(Yes, as far as I remember the job shell will need to use a ptrace or other call to stop in `main()` and wait for debugger attach)

**Comment by dongahn** [comment]
At some point, having a step-by-step description on how our dev tools, in particular debuggers will be supported under S4. We not only need this sync capability but also a process that the debugger front end will attach to and use the binary MPIR debug interface.

**Comment by grondo** [comment]
@dongahn, MPIR debug interface is easy to implement, but I thought perhaps we had developed something more scalable using KVS in our proof-of-concept we did early on?

**Comment by dongahn** [comment]
Yes, that will be the new capabilities we will develop to provide a better interoperability + scalability among job elements etc. And tools developers including myself can progressively take advantages of these capabilities. 

But we definitely need to support MPIR debug interface etc for legacy tools. I don't have full control over what third party tools will develop or adopt Flux-only solutions. 

BTW, on the new capabilities -- at some point we need to reach out to stake holders and ppl who desire to standardize these things. Right now the projects to look into: 1) MPI Tools subcommittee and 2) also PMIX.

**Comment by grondo** [comment]
Thanks, @dongahn.

I think block diagram or very simple flowchart of a job shell implementation is a good idea. I'll bootstrap that activity. I'm hopeful that MPIR_proctable could be just one of many plugins that are dynamically loaded by job shells as needed.

**Comment by dongahn** [comment]
Sounds like a plan. I believe I can be a reasonable reviewer for the MPIR part :-) I think one of the tricky parts will be support for MPIR_partial_attach_ok. I will feed more info when appropriate.
